src/js/shot-generator-new/hooks/use-assets-manager.js

```diff
@@ -1,101 +1,172 @@
 import * as THREE from 'three'
-import React, { useState, useReducer, useMemo, useCallback } from 'react'
+import React, { useState, useMemo, useEffect, useRef } from 'react'
 import '../../vendor/three/examples/js/loaders/GLTFLoader'
+import observable from "../../utils/observable"
 
-const reducer = (state, action) => {
-  const { type, payload } = action
-  const { id, progress, value, error } = payload
-
-  switch (type) {
-    case 'PENDING':
-      // ignore if already exists
-      return (state[id])
-        ? state
-        : {
-          ...state,
-          [id]: { status: 'NotAsked' }
-        }
-    case 'LOAD':
-      // ignore if already loading
-      return (state[id].loading)
-        ? state
-        : {
-          ...state,
-          [id]: { status: 'Loading' }
+/**
+ * Resources storage
+ * @type {observable}
+ */
+export const cache = observable({})
+
+const gtlfLoader = new THREE.GLTFLoader()
+const textureLoader = new THREE.TextureLoader()
+
+const LOADING_MODE = {
+  PENDING: 'PENDING',
+  SUCCESS: 'SUCCESS',
+  ERROR: 'ERROR'
+}
+
+/**
+ * Checks if resource has been loaded
+ * @param path Resource path
+ * @returns {boolean}
+ */
+const isLoaded = (path) => {
+  const state = cache.get()
+
+  return (state[path] && state[path].status === LOADING_MODE.SUCCESS)
+}
+
+/**
+ * Checks if asset is loaded or if asset fetching has been started
+ * @param path
+ * @returns {boolean}
+ */
+const assetExist = (path) => Boolean(cache.get()[path])
+
+/**
+ * Fetches resource
+ * @param path Resource path
+ * @returns {Promise<null|THREE.Texture|THREE.Object>}
+ */
+export const loadAsset = (path) => {
+  if (!path) {
+    return Promise.resolve(null)
+  }
+
+  const current = cache.get()
+
+  if (!current[path]) {
+    cache.set({
+      ...cache.get(),
+      [path]: {data: null, status: LOADING_MODE.PENDING}
+    })
+
+    return new Promise((resolve, reject) => {
+      const current = cache.get()
+
+      if (!current[path].data) {
```

Comment:
if .data is null, but loading has already started, this will still trigger, and try to call .load again. Is that a problem? Using the RemoteData pattern (initializing with a "NotAsked" state and checking for that before transition to a "Pending" state) could solve this.